File: CS/DataContextSLSample/ViewClasses.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Controls;
using System.Windows;
using System.Windows.Data;

namespace DataContextSLSample {

    public class SomeVisualControl : Control, IDataContextOwner {
        DataContextBinder dataContextBinder;

        public SomeVisualControl() {
            this.dataContextBinder = new DataContextBinder(this);
        }
        public static readonly DependencyProperty InnerDataStoreProperty =
            DependencyProperty.Register("InnerDataStore", typeof(DataStore), typeof(SomeVisualControl), new PropertyMetadata(null));

        public DataStore InnerDataStore { 
            get { return (DataStore)GetValue(InnerDataStoreProperty); } 
            set { SetValue(InnerDataStoreProperty, value); } 
        }


        object IDataContextOwner.DataContext {
            get { return DataContext; }
        }
        void IDataContextOwner.UpdateInnerDataContext(object dataContext) {
            if (InnerDataStore != null)
                InnerDataStore.DataContext = dataContext;
        }
    }
    
}

File: VB/DataContextSLSample/ViewClasses.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows.Controls
Imports System.Windows
Imports System.Windows.Data

Namespace DataContextSLSample

	Public Class SomeVisualControl
		Inherits Control
		Implements IDataContextOwner
		Private dataContextBinder As DataContextBinder

		Public Sub New()
			Me.dataContextBinder = New DataContextBinder(Me)
		End Sub
		Public Shared ReadOnly InnerDataStoreProperty As DependencyProperty = DependencyProperty.Register("InnerDataStore", GetType(DataStore), GetType(SomeVisualControl), New PropertyMetadata(Nothing))

		Public Property InnerDataStore() As DataStore
			Get
				Return CType(GetValue(InnerDataStoreProperty), DataStore)
			End Get
			Set(ByVal value As DataStore)
				SetValue(InnerDataStoreProperty, value)
			End Set
		End Property


		Private ReadOnly Property IDataContextOwner_DataContext() As Object Implements IDataContextOwner.DataContext
			Get
				Return DataContext
			End Get
		End Property
		Private Sub UpdateInnerDataContext(ByVal dataContext As Object) Implements IDataContextOwner.UpdateInnerDataContext
			If InnerDataStore IsNot Nothing Then
				InnerDataStore.DataContext = dataContext
			End If
		End Sub
	End Class

End Namespace
